diff --git a/src/a-b-c/a-b-c.d/19-config-adds b/src/a-b-c/a-b-c.d/19-config-adds
index 6668c5b..fd25b16 100644
--- a/src/a-b-c/a-b-c.d/19-config-adds
+++ b/src/a-b-c/a-b-c.d/19-config-adds
@@ -18,7 +18,7 @@ c_add midway 'run config_calxeda; u_dtb=ecx-2000.dtb'
 
 #Freescale i.MX such as Wandboard and Utilite
 c_add imx 'run config_default; u_extraargs="\$u_extraargs console=\$console,115200"; u_k_addr=0x14000000; u_ramfs_addr=0x32000000; u_dtb_addr=0x12000000; skip_dtb=""; pass_dtb=1; u_iodevs="mmc usb"; u_fs=ext2; u_boot=bootz; u_devpart="0:1 0:2 1:1 1:2"'
-c_add utilite 'run config_imx; u_boot=bootm; skip_dtb=1; pass_dtb=""; u_iodevs="sata mmc usb"'
+c_add utilite 'run config_imx; u_dtb="imx6q-cm-fx6.dtb"; u_iodevs="sata mmc usb"; u_devpart="0:1 0:2 1:1 1:2 2:1 2:2"'
 c_add wandboard 'run config_imx; setenv u_dtb \$fdt_file'
 c_add cubox 'run config_imx; setenv u_dtb \$fdt_file'
 
diff --git a/src/a-b-c/a-b-c.d/21-autodetect b/src/a-b-c/a-b-c.d/21-autodetect
index 41097de..0283b65 100644
--- a/src/a-b-c/a-b-c.d/21-autodetect
+++ b/src/a-b-c/a-b-c.d/21-autodetect
@@ -29,6 +29,10 @@ elif test -n \$fdt_file; then
   if test \$fdt_file = imx6q-wandboard.dtb; then
     run config_wandboard
   fi
+elif test -n \$fdtfile; then
+  if test \$fdtfile = cm-fx6.dtb; then
+    run config_utilite
+  fi
 elif test -n \$ethact; then
   if test \$ethact = asx0; then
     run config_arndale
diff --git a/src/a-b-c/a-b-c.d/85-static-routines b/src/a-b-c/a-b-c.d/85-static-routines
index d46b198..c982d08 100644
--- a/src/a-b-c/a-b-c.d/85-static-routines
+++ b/src/a-b-c/a-b-c.d/85-static-routines
@@ -6,7 +6,7 @@ cat << EOF
 setenv showdefs 'echo ""; echo "The following configuration values are defined:"; echo u_k_addr: \$u_k_addr; echo u_k_addr: \$u_k_addr; echo u_ramfs_addr: \$u_ramfs_addr; echo u_dtb_addr: \$u_dtb_addr; echo u_iodevs: \$u_iodevs; echo u_iodevs: \$u_iodevs; echo u_fs: \$u_fs; echo u_devpart: \$u_devpart; echo u_boot: \$u_boot; echo u_root: \$u_root; echo u_dtb: \$u_dtb; echo skip_dtb: \$skip_dtb (null means a dtb will be loaded); echo pass_dtb: \$pass_dtb (null means no bootarg passed); echo u_extraargs: \$u_extraargs'
 
 # catcat: variable pointer conversion: input catin, output catout
-setenv catcat setenv catout\;'setenv catX "setenv catout '\\\\\\\\\\\\\''\\$\\\$catin'\\\\\\\\\\\\\\''"' \; run catX
+setenv catcat setenv catout\;'setenv catX setenv catout '\\\\\\\\\\\\\\"'\\\$\\\$catin'\\\\\\\\\\\\\\"'' \; run catX
 
 # Something failed.
 setenv failed 'echo The last operation failed.; run help; run false'
